add how nodes work to how swarm works guide

sfsmithcha opened the pull request:
Add first topic to "How swarm mode works" guide: How nodes work.

Signed-off-by: Charles Smith charles.smith@docker.com

sfsmithcha committed:
add how nodes work to how swarm works guide

Signed-off-by: Charles Smith <charles.smith@docker.com>

sfsmithcha commented:
ping @mgoelzer . separating out the individual topics in this guide for smaller prs

abronan commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+manager nodes.
+* In general, an `N` manager cluster will tolerate the loss of at most
+`N/2` managers.
+* Seven is the maximum possible number of manager nodes for a swarm.
```
I don't think this sentence is correct: 7 is the maximum **recommended** number. But you can go beyond if you want provided that you don't mind the performance cost.

sfsmithcha replied:
thanks, this was a comment from @mgoelzer

mgoelzer replied:
Yep, was my mistake -- thanks @abronan !

abronan commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+
+## Changing roles
+
+You can promote a worker node to be a manager manager by running `docker node promote`.
```
redundant `manager`

abronan commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+## Changing roles
+
+You can promote a worker node to be a manager manager by running `docker node promote`.
+For example, if you may want to promote a manager node when you to
```
`you may want to promote a manager node if you want to <...>`?

thaJeztah commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+
+Using a [Raft](https://raft.github.io/raft.pdf) implementation, the
+managers maintain a consistent internal state of the entire swarm and
+all the services running on it. In most cases, one manager should be
```
I'm not sure we should say "most cases". I think a single manager is only suitable for testing / really small scale setups

mgoelzer replied:
Agree on remove most cases, too prescriptive, but I do think we should convey that single manager is the correct decision if you don't care about HA.

thaJeztah replied:
My point is, basically, that running Swarm mode with a single manager is actually
*less* HA than running a bunch of nodes without Swarm mode. If you have a single
manager, and that manager goes down (or looses connection with the workers), you
cannot do anything (sure, your containers/services keep running).

With a "non-swarm-mode" setup, you'd still be able to manually start/stop/create
containers, but since services are "managed", interacting with those containers
can be troublesome (when the manager comes back, strange things happen). Also
(but this may be fixed).

In a more serious situation, when the manager goes down *permanently*, your cluster
is a write-off; you cannot recover from that, without creating a new cluster.

We introduced Swarm mode, pointing out that there's no single point of failure,
and that's true from a design perspective (distributed store, etc.), and when
set up with multiple managers; but (unless I'm missing something), with a single
manager, the manager is the risk.

For that reason, even for a small cluster, I'm not sure we can ever recommend
running with a single manager, for anything that's "production".

I think we should point out these risks; and explain that a single manager is only
useful for testing, or other situations where loosing the swarm is no problem.

abronan replied:
Agreed with the comment above, we should remove that sentence. Single Manager is only suitable for testing. But it's up to the user to decide what is reasonable or not for their use case.

abronan commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+* A five-manager swarm tolerates a maximum simultaneous loss two
+manager nodes.
+* In general, an `N` manager cluster will tolerate the loss of at most
+`N/2` managers.
```
`(N-1)/2` managers

abronan commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+* A three-manager swarm tolerates a maximum loss of one manager.
+* A five-manager swarm tolerates a maximum simultaneous loss two
+manager nodes.
+* In general, an `N` manager cluster will tolerate the loss of at most
```
remove `In general`

thaJeztah commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+
+    >**Important Note**: Adding more managers does NOT mean increased
+    scalability or higher performance. In general, the opposite is
+    true. The only reason to run more than one manager in a swarm is
```
I'd remove `"The only reason to.."` as it may be interpreted as "You really shouldn't add more than one manager, unless you really need to".

Sentence may need to be slightly rewritten after removing that of course.

thaJeztah commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+create` and the scheduler will place all tasks on the local Engine.
+
+To prevent the scheduler from placing tasks on a manager node in a multi-node
+swarm, you can put the manager node into `Drain` availability. The scheduler
```
s/you can put/you put/

thaJeztah commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+`Active` node. The scheduler does not assign new tasks to nodes with `Drain`
+availability.
+
+See the command line documentation for [`docker node update`](../../reference/commandline/node_update.md).
```
s/See/Refer to/ ?

Perhaps;

```
Refer to the `docker node update` command line reference for
changing the availability of a node
```

thaJeztah commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+You can also demote a manager node to a worker node. See
+[node demote](../../reference/commandline/node_demote.md).
+
+<!-- For when How services work guide is ready
```
nit: can you add `TODO (sfsmitcha)` here?

thaJeztah commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,94 @@
+five manager nodes according to your organization's high-availability
+requirements:
+
+* A three-manager swarm tolerates a maximum loss of one manager.
```
Do we need to explicitly warn against using *two* managers? As that *decreases* fault tolerance (a single manager down means you can no longer manage the swarm)

sfsmithcha replied:
I thought 2 managers was not _decreasing_ fault tolerance. It is just the same as running 1 manager. Could call that out even though we already say to run just 1.

thaJeztah replied:
Well, actually it it's twice as risky; with two managers, chance that one fails is twice as large, but the same result as a single manager (and it being down), because with one of two managers down, consensus will never be reached, so the manager that's still up is not allowed to make changes.

sfsmithcha commented:
@thaJeztah , @abronan PTAL. I think I addressed the concerns here. Once the admin guide  https://github.com/docker/docker/pull/24857/ is ready,  I can link there from the manager nodes. This shows that 2 managers doesn't increase the fault tolerance.

Thanks again for the great feedback & help.

thaJeztah commented:
@sfsmithcha did you push your changes?

sfsmithcha commented:
@thaJeztah pushed now.

thaJeztah commented on docs/swarm/how-swarm-mode-works/nodes.md:
```diff
@@ -0,0 +1,93 @@
+availability.
+
+Refer to the [`docker node update`](../../reference/commandline/node_update.md)
+command line reference to see how to change node availabily
```
oops; s/availabily/availability./

thaJeztah replied:
feel free to fix that in the follow up

thaJeztah commented:
Thanks @sfsmithcha LGTM

go ahead and merge if you want

vdemeester commented:
LGTM 👼
